gwcs/spectroscopy.py

```diff
@@ -192,3 +196,223 @@ def input_units(self):
             return {'wavelength': 1 / self.groove_density.unit,
                     'alpha_in': u.Unit(1),
                     'beta_in': u.Unit(1)}
+
+
+class Snell3D(Model):
+    """
+    Snell model in 3D form.
+
+    Inputs are direction cosines.
+
+    Parameters
+    ----------
+    n : float
+        Refractive index of the material.
+
+    Returns
+    -------
+    alpha_out, beta_out, gamma_out : float
+        Direction cosines.
+    """
+    _separable = False
+    linear = False
+
+    n_inputs = 3
+    n_outputs = 3
+
+    n = Parameter(default=1)
+
+    def __init__(self, n=n, **kwargs):
+        super().__init__(n=n, **kwargs)
+        self.inputs = ('alpha_in', 'beta_in', 'gamma_in')
+        self.outputs = ('alpha_out', 'beta_out', 'gamma_out')
+
+    @staticmethod
+    def evaluate(alpha_in, beta_in, gamma_in, n):
+        # Apply Snell's law through front surface,
+        # eq 5.3.3 II in Nirspec docs
+        xout = alpha_in / n
+        yout = beta_in / n
+        zout = np.sqrt(1.0 - xout**2 - yout**2)
+        return xout, yout, zout
+
+
+class SellmeierGlass(Model):
```

Comment:
I take it that this is there just to be complete and that it actually isn't used for NIRSPEC (instead the SellmeierZemax is used instead)?

Reply:
They are both used depending on the value of `T - Tref`.  They expect `SellmeierGlass` to be used typically ( when `Tdiff < 20 K`) but don't exclude the possibility of going into the other regime. At least this is the information I have.